### ISSUE TITLE
Handling the registry.json in the Jenkins CI context

### ISSUE THREAD

**Issue by petermbauer** [conan-io/conan#5510]
Hi,

here is what i am trying to do with the Jenkins CI:
- use the Artifactory credentials configured on the Jenkins Master (Artifactory Plugin) for configuring a respective Conan remote. This means the name of the remote changes with every build
- re-use the Conan local cache between the jobs to avoid re-downloading the packages

Therefore the following steps are executed:
```
// in case there was a registry.json with remotes in the previously installed Conan config
bat "conan remote clean"
// get the credentials from the plugin config
artifactoryServer = Artifactory.server('myserverconfigname')
conanClient = Artifactory.newConanClient()
// use the credentials to add a remote, conanRemote holds the generated name of the new remote
conanRemote = conanClient.remote.add server: artifactoryServer, repo: 'myrepo'
```

My question is if this can lead to errors like this
```
ERROR: Error while trying to get recipe sources for conan-commonlib/2.0.0@myuser/mychannel. No remote defined
```
when running e.g. `conan create --update --remote d541d675-05a3-4ec0-a80f-58e165f91c71 . myuser/mychannel`
`conan-commonlib` is a package used with `python_requires()`

What is the recommended way of dealing with the `registry.json` in such a setup? Is it even possible to keep the local cache between the builds?

We are using Conan 1.17.0 under Windows 10.

thx!

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by jgsogo** [comment]
Hi, @petermbauer. As long as you know what you are doing, you can reuse the local cache between jobs, although for a general CI I'd say it is not recommended. You may end using binaries from dependencies that are not in your server and get non-reproducible builds, concurrency issues,... so my recommendation is to go always through the server.

In your use case, you can try to rename the remote instead of cleaning them. You can just `conan remote rename` or you can iterate your packages setting a different remote for them with `conan remote add_ref/add_pref`.

Nevertheless, I'm going to take a closer look at the ERROR you report, it may be an actual bug.

**Comment by jgsogo** [comment]
About the ERROR:
 * In your cache you managed to get the recipe and the binaries of the library `conan-commonlib/2.0.0@myuser/mychannel` without building it locally (probably it was available in the remote). The keypoint is that you don't have your sources in the cache.
 * You clean the remote associated with `conan-commonlib/2.0.0@myuser/mychannel`
 * You consume it as a python-require, so Conan needs to retrieve the sources (files declared in `ConanFile::exports_sources` member attribute). Reusing sources is documented [here](https://docs.conan.io/en/latest/extending/python_requires.html#reuse-source-files)
 * ...but as you do not have a `remote` defined it fails, Conan cannot get the sources.

It fails with a python_requires, but it would fail with any package:
 * `conan remove zlib/1.2.11@conan/stable -f`
 * `conan install zlib/1.2.11@conan/stable`
 * `conan remote remove_ref zlib/1.2.11@conan/stable`
 * `conan install zlib/1.2.11@conan/stable --build zlib`

So yes, the error you are getting is expected.

**Comment by jgsogo** [comment]
Hi, @petermbauer 

I've been having a look at the Artifactory Jenkins plugin and now I understand better your problem and what you are trying to do. Given the current version of the plugin you do not have any alternative, but I'm proposing this PR (https://github.com/jfrog/jenkins-artifactory-plugin/pull/177) to the plugin repository.

If you are reusing the cache between several jobs, then you shouldn't clean the remotes, you just need to call `conan user <username> -p <password> -r <remote>` to authenticate and be able to upload your artifacts to the configured remote using the credentials from the Artifactory plugin: you need to call `conan user` withoug calling `conan remote add` which is something you cannot do right now.

If that PR gets accepted you will be able to:
 * share the cache between your jobs (remotes will maintain the names), or call `conan config install <url>` to get a shared configuration for your team
 * authenticate to that remote (see example in the PR)
 * create and upload the artifacts

Once again, I advise you to be careful when sharing the cache, you can have race conditions, reuse binaries from the cache instead of downloading them from the server,... 

Let me know if this would be a solution for you and we will try to push forward that PR.

Thanks

**Comment by petermbauer** [comment]
Hi @jgsogo, thank you very much for the great support and your efforts!

Since some packages (e.g. Qt) tend to be quite big (~3GB) due to e.g. debug files its not really an option to install them for each build.

I was also considering having a Conan local cache per Jenkins workspace/job which would avoid the race conditions when using multiple executors and possible incompatibilities between the Conan versions but then it would not be possible to let Jenkins clean the workspace before each build, leading to other potential issues.

The change to the plugin will for sure improve things in any case. In the meantime i will try to add a `conan remote rename` after cleaning all remotes and `conanClient.remote.add` to maintain the same name for the remote. I hope i understood that correctly?

**Comment by jgsogo** [comment]
The problem after cleaning all remotes is that the association between a reference and a remote will be broken and you won't be able to look for updates of that reference or to download binaries...

It might be ok not to be able to download new versions for packages already in the cache: if they are in the cache (and the cache is shared) it is because they have been built in another job (or downloaded).

So, the workaround is a little bit tricky, but I think it could work. The steps to follow should be:
1. Clean all the remotes
2. Add the remote corresponding to your Artifactory, keep the `remoteName`
3. Build your package, test,...
5. Upload: `conan upload <your-ref> -r remoteName

Problems:
 * if these steps are being run in several jobs in parallel you will eventually get a race condition
 * if you need a new version of a reference that is already in the cache, as the reference is not tracking any remote, it will fail (you will need a special job for this update).

I'm afraid there is no good solution right now, any alternative that comes to my mind has some drawback... unless you directly call the command line to be able to run any Conan command with any argument.

This is one of the reason we will try to have a closer look at the plugin and evolve it.

**Comment by petermbauer** [comment]
> The problem after cleaning all remotes is that the association between a reference and a remote will be broken and you won't be able to look for updates of that reference or to download binaries...

After trying a few things out i found the following workaround:
1. `conan remote list` -> get name and full URL of remote with URL ending with the Artifactory repo name "myrepo". This is the existing remote referenced in the `registry.json`
    1. if one is found: `conan remote update <existing remote name> <dummy URL>` to avoid error when adding the new remote with same URL with the Jenkins plugin
1. `conanClient.remote.add server: artifactoryServer, repo: "myrepo"`
If there was no existing remote found before we are already done and continue with the remote added by the plugin. Otherwise we need to continue using the old remote to maintain the references in the registry:
1. `conan remote list` -> get full URL of the new UUID remote
1. `conan remote remove <new UUID remote>` -> get rid of new remote but the authentication token is stored per URL and not per remote name so it is kept in the `.conan.db`
1. `conan remote update <existing remote name> <full URL>` -> switch the existing remote back to the original URL so it can be used further on but with the authentication provided by the Jenkins plugin.

**Comment by jgsogo** [comment]
Although your solution works, there a couple of comments or TODOs I keep for myself:
 * it requires calling the command line and parsing the output (`conan remote list`), this functionality must be incorporated to the plugin if it is really needed
 * The behavior you mention in step `4`, I think it is more a _bug_ than a _feature_, probably if a remote is removed the authentication should be removed too

Also, there is an intrinsic problem if several jobs are using the same cache: one of them could be executing step `1` (setting a dummy URL to a remote) and another job may be using the URL to retrieve binaries... 

I'll try to come up with a solution (adding functionality to the plugin) and I will share it here: https://github.com/conan-io/conan/issues/5553

Thanks for sharing.

**Comment by jgsogo** [comment]
Hi, @petermbauer. I've opened this PR (https://github.com/jfrog/jenkins-artifactory-plugin/pull/179) to the Artifactory plugin, I think it will fit your use case 😉 

With the `force` argument (translated into `conan remote add --force ...`) you are able to add the credentials to your remote while keeping packages in the cache associated with that remote. I really think this is the way to go as soon as it is released.

**Comment by jgsogo** [comment]
Yes, please, leave it open until that PR is released. I know it will be duplicated, but I'll take care of it 😉

**Comment by jgsogo** [comment]
Hi! @petermbauer . The PR has been merged and the new release ([v3.4.0](https://plugins.jenkins.io/artifactory)) contains these changes. 

The function `client.remote.add` now accepts arguments `remoteName` and `force`, an example would be:
 1. First run `conan config install <repository/zip>`. It contains a remote called `myremote`.
 2. Use the credentials from the Artifactory/Jenkins plugin:
    ```groovy
    def artifactory_repo = "conan-local"
    def artifactory_name = "Artifactory Docker"

    def server = Artifactory.server artifactory_name
    def client = Artifactory.newConanClient()
    client.remote.add server: server, repo: artifactory_repo, remoteName: "myremote", force: true
    ```
    It will translate to the following commands:
      *  `conan remote add --force myremote <url>`
      *  `conan user -p <password> -r myremote <username>` (invisible command)

**Comment by jgsogo** [comment]
Feel free to open it again if it is not working properly. Thanks!
